src/groupeddataframe/splitapplycombine.jl

```diff
@@ -597,33 +606,55 @@ function combine_helper(f, gd::GroupedDataFrame,
         else
             newparent = parent(gd)[idx, gd.cols]
         end
-        hcat!(newparent, select(valscat, Not(intersect(keys, _names(valscat))), copycols=false),
+        hcat!(newparent,
+              select(valscat, Not(intersect(keys, _names(valscat))), copycols=false),
               copycols=false)
         ungroup && return newparent
 
-        if length(idx) == 0
+        if length(idx) == 0 && !(keeprows && length(keys) > 0)
             @assert nrow(newparent) == 0
             return GroupedDataFrame(newparent, collect(1:length(gd.cols)), Int[],
-                                    Int[], Int[], Int[], 0, Dict{Any,Int}())
+                                    Int[], Int[], Int[], 0, Dict{Any,Int}(),
+                                    Threads.ReentrantLock())
         end
         if keeprows
+            @assert length(keys) > 0 || idx == gd.idx
+            @assert names(newparent, 1:length(gd.cols)) == names(parent(gd), gd.cols)
             # in this case we are sure that the result GroupedDataFrame has the
-            # same structure as the source
-            # we do not copy data as it should be safe - we never mutate fields of gd
```

Comment:
When do we mutate fields now?

Reply:
we do not mutate fields now.
It is a preparation for `filter!`. But I prefer to make a copy here as it is not very expensive, and be safe that in the future changing if someone introduces mutation to `GroupedDataFrame` it does not get forgotten (as it is easy to forget that this line in such a large codebase assumes that it is not mutated).

Reply:
I'd rather avoid making copies preventively as long as we never mutate fields. Especially if we drop `filter!` for now.

Reply:
OK - changed. I have made an extensive implementation note, as `GroupedDataFrame` is currently quite tricky to implement new functionality for.